col/src/main/java/vct/col/ast/util/NameScanner.scala

```diff
@@ -36,14 +34,42 @@ class NameScanner extends RecursiveVisitor[AnyRef](null, null) {
 
   val frameStack: mutable.Stack[mutable.Set[DeclarationStatement]] = mutable.Stack()
 
-  // I suspect there is some external code that depends on insertion order... Therefore use LinkedHashMap
+  /**
+    * Given/yields must be stacks as well, in case with/then blocks are nested
+    */
+  val givenStack: mutable.Stack[mutable.Set[DeclarationStatement]] = mutable.Stack()
+
+  val yieldsStack: mutable.Stack[mutable.Set[DeclarationStatement]] = mutable.Stack()
+
+  /**
+    * There might be external code that depends on insertion order. Therefore use LinkedHashMap
+    */
   val freeNames: mutable.Map[String, Entry] = mutable.LinkedHashMap()
 
   // Ensure there is at least one frame at the start
-  push()
+  pushFrame()
 
-  private def hasDecl(name: String): Boolean = frameStack.exists(frame => frame.exists(p => p.name == name))
+  /**
+    * Finds the first declaration in the framestack that has name `name`
+    */
+  private def getDecl(name: String): Option[DeclarationStatement] = frameStack
+    // Unlift turns a function returning Option into a partial function
+    .collectFirst(Function.unlift(_.find(decl => decl.name == name)))
```

Comment:
While this may be very nice functionally, I don't think it's very readable (I still had to look up what exactly a partial function in scala is etc.). Maybe you could consider:

```
frameStack.collectFirst(frame => frame.find(_.name == name) match {
    case Some(decl) => decl
})
```

I think that still produces a partial function with respect to the match, or maybe even:

```
frameStack.map(_.find(_.name == name)).collectFirst { case Some(decl) => decl }
```

Reply:
I could not resist making an unreadable functional implementation:

```
frameStack.map(_.find(_.name == name)).reduce(_ orElse _)
```

Reply:
or even `frameStack.map(_.find(_.name == name)).flatten.headOption`, so many options!

Reply:
more seriously, I think scala docs imply that the order of iterating over a stack is FIFO, I don't think that is what you want?

Reply:
The main idea for using unlift is to prevent looking in every frame of the stack. However I agree it's a bit cryptic, it's probably better to change it to `frameStack.map` and eat the performance penalty. (I think your very first example is the most clear, but since it starts with `frame => ...` it is still a regular (total?) function and not a partial function, so it wouldn't work I think. But I'll doublecheck it)

The iteration starts at the top of the stack, which means the most recent declaration is found first if I understand correctly. I think that is what we want here.